packages/dialog/src/vaadin-dialog.js

```diff
@@ -53,18 +86,200 @@ export class DialogOverlay extends OverlayElement {
+      const footerContainer = this.footerContainer;
+      footerContainer.innerHTML = '';
+      // Whenever a Lit-based renderer is used, it assigns a Lit part to the node it was rendered into.
+      // When clearing the rendered content, this part needs to be manually disposed of.
+      // Otherwise, using a Lit-based renderer on the same node will throw an exception or render nothing afterward.
+      delete footerContainer._$litPart$;
+    }
+
+    if (
+      (headerRenderer && (headerRendererChanged || openedChanged)) ||
+      (footerRenderer && (footerRendererChanged || openedChanged))
+    ) {
+      if (opened) {
+        this.requestContentUpdate();
+      }
+    }
+    this._toggleHasHeaderAttribute();
+    this._toggleHasFooterAttribute();
+  }
+
+  _toggleHasHeaderAttribute() {
+    if (this.headerRenderer) {
+      this.setAttribute('has-header', '');
+    } else {
+      this.removeAttribute('has-header');
+    }
+  }
+
+  _toggleHasFooterAttribute() {
+    if (this.footerRenderer) {
+      this.setAttribute('has-footer', '');
+    } else {
+      this.removeAttribute('has-footer');
+    }
+  }
+
+  _headerTitleChanged(headerTitle, opened) {
+    const headerTitleChanged = this._oldHeaderTitle !== headerTitle;
+    this._oldHeaderTitle = headerTitle;
+
+    const openedChanged = this._oldOpened !== opened;
+    this._oldOpened = opened;
+
+    if (headerTitleChanged || openedChanged) {
+      if (this.opened) {
+        this.requestContentUpdate();
+      }
+    }
+    this._toggleHasTitleAttribute();
+  }
+
+  _toggleHasTitleAttribute() {
+    if (this.headerTitle) {
+      this.setAttribute('has-title', '');
+    } else {
+      this.removeAttribute('has-title');
+    }
+  }
+
+  _headerTitleRenderer() {
+    const headerTitleElement = this.getHeaderTitleElement();
+    if (this.headerTitle) {
+      headerTitleElement.style.display = '';
+      headerTitleElement.textContent = this.headerTitle;
+    } else {
+      headerTitleElement.style.display = 'none';
+    }
+  }
+
+  /**
+   *
+   * @returns HTMLElement header title element
+   */
+  getHeaderTitleElement() {
+    let headerTitleElement = this.querySelector('[slot=title]');
+
+    if (!headerTitleElement) {
+      headerTitleElement = document.createElement('span');
+      headerTitleElement.id = 'title';
+      headerTitleElement.setAttribute('slot', 'title');
+      this.setAttribute('aria-labelledby', 'title');
+      this.appendChild(headerTitleElement);
+    }
+
+    return headerTitleElement;
+  }
+
+  requestContentUpdate() {
+    super.requestContentUpdate();
+
+    if (this.headerRenderer) {
+      this.headerRenderer.call(this.owner, this.headerContainer);
+    }
+
+    if (this.footerRenderer) {
+      this.footerRenderer.call(this.owner, this.footerContainer);
+    }
+
+    this._headerTitleRenderer();
+  }
+
+  get headerContainer() {
+    let headerContainer = this.querySelector('vaadin-dialog-header');
+    if (!headerContainer) {
```

Comment:
We probably should move this logic outside of the getter. IMO it's better to create container when `headerRenderer` is set, and remove it when `headerRenderer` is removed.

Then `headerContainer` would only return an element when `headerRenderer` is defined.